File: java/src/CanSum.java
import java.util.ArrayList;
import java.util.*;

public class CanSum {

    private static Map<Integer, Boolean> resultMap = new HashMap<>();

    public static void main(String[] args) {
        System.out.println("This is my can sum program..");
        int sum = 7000;
        boolean result = canSum(sum, new int[] { 5, 3, 4, 7 });
        System.out.println("Result for sum [" + sum + "] -> " + result);
    }

    private static boolean canSum(int sum, int[] args) {

        if (resultMap.containsKey(sum))
            return resultMap.get(sum);

        if (sum == 0)
            return true;
        if (sum < 0)
            return false;

        for (int key : args) {
            int diff = sum - key;
            boolean result = canSum(diff, args);
            if (result) {
                resultMap.put(sum, result);
                return true;
            }
        }

        resultMap.put(sum, false);
        return false;
    }
}

File: python/cansum.py

import sys
sys.setrecursionlimit(10000)

def canSum(sum, args):

    if sum in data.keys():
        return data[sum];

    if sum == 0:
        return 1    

    if sum < 0:
        return 0    

    for key in args: 
        diff = sum - key;
        if canSum(diff, args) == 1:
            data[sum] = 1;
            return 1;
        
    data[sum] = 0
    return 0


print("This is my can sum program to learning Dynamic Programming....")
sum = 7000

data = dict()
result = canSum(sum, [5, 3, 4, 7])
print("Result for sum [" + str(sum) + "] -> " + str(result))